boa/src/exec/tests.rs

```diff
@@ -1235,3 +1235,135 @@ fn not_a_function() {
     "#;
     assert_eq!(forward(&mut engine, scenario), "TypeError: not a function");
 }
+
+#[test]
+fn assignment_to_non_assignable() {
+    // Relates to the behaviour described at
+    // https://tc39.es/ecma262/#sec-assignment-operators-static-semantics-early-errors
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    const ERR_MSG: &str =
+        "Parsing Error: Syntax Error: Syntax Error: Invalid left-hand side in assignment";
+
+    // Tests all assignment operators as per [spec] and [mdn]
+    //
+    // [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_Operators#Assignment
+    // [spec]: https://tc39.es/ecma262/#prod-AssignmentOperator
+    assert_eq!(forward(&mut engine, "3 = 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 += 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 -= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 *= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 /= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 %= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 &= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 ^= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 |= 5"), ERR_MSG);
+}
+
+#[test]
+fn multicharacter_assignment_to_non_assignable() {
+    // Relates to the behaviour described at
+    // https://tc39.es/ecma262/#sec-assignment-operators-static-semantics-early-errors
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    const ERR_MSG: &str =
+        "Parsing Error: Syntax Error: Syntax Error: Invalid left-hand side in assignment";
+
+    assert_eq!(forward(&mut engine, "3 **= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 <<= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 >>= 5"), ERR_MSG);
+}
+
+#[test]
+#[ignore]
+fn multicharacter_bitwise_assignment_to_non_assignable() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    const ERR_MSG: &str =
+        "Parsing Error: Syntax Error: Syntax Error: Invalid left-hand side in assignment";
+
+    // Disabled - awaiting implementation.
+    assert_eq!(forward(&mut engine, "3 >>>= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 &&= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 ||= 5"), ERR_MSG);
+    assert_eq!(forward(&mut engine, "3 ??= 5"), ERR_MSG);
+}
+
+#[test]
+fn assignment_line_terminator() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    let init = r#"
+    let a = 3;
+
+    a = 
+    5;
+
+    a
+    "#;
+
+    assert_eq!(forward(&mut engine, init), "5");
+}
+
+#[test]
+fn assignment_multiline_terminator() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    let init = r#"
+    let a = 3;
+
+
+
+
+
+    a = 
+
+
+
+
+
+
+
+    
+    5;
+
+
+
+
+
+
+    a
+    "#;
```

Comment:
Hmmm. this to me seems like we are testing the lexer/parser rather than the execution behaviour

Reply:
Good point, will move